include/intel/cpu_support.h

```diff
@@ -0,0 +1,384 @@
+//  std::endl; }
+#ifndef log_intel
+#define log_intel(x)                  \
+  if (IntelKernel<>::log_enabled()) { \
+    LOG(INFO) << x;                   \
+  }
+#endif
+
+static inline Xbyak::Zmm make_zmm(const Xbyak::Xmm &v) {
+  return Xbyak::Zmm(v.getIdx());
+}
+template <int version = 0>
+struct IntelKernel {
+  static int64_t getValue() {
+    int64_t v = 0;
+    const char *label = "DGL_CPU_INTEL_KERNEL_ENABLED";
+    const char *ptr = std::getenv(label);
+    if (ptr) {
+      v = atoll(ptr);
+      log_intel(label << "=>" << v);
+    }
+    return v;
+  }
+
+  static int64_t enabled() {
+    static int64_t r = IntelKernel<version>::getValue();
+    return r;
+  }
+
+  static int log_enabled() {
+    static int r = (std::getenv("DGL_CPU_INTEL_KERNEL_LOG")) ? 1 : 0;
+    return r;
+  }
+};
+
+namespace op_ns = ::dgl::aten::cpu::op;
+
+template <typename T, typename Tuple>
+struct has_type;
+
+template <typename T>
+struct has_type<T, std::tuple<>> : std::false_type {};
+
+template <typename T, typename U, typename... Ts>
+struct has_type<T, std::tuple<U, Ts...>> : has_type<T, std::tuple<Ts...>> {};
+
+template <typename T, typename... Ts>
+struct has_type<T, std::tuple<T, Ts...>> : std::true_type {};
+
+typedef std::tuple<float, double> supported_types;
+
+template <class OCmp, template <class> class ToP, class Tup,
+          int ok = std::tuple_size<Tup>::value>
+struct DeepType;
+
+template <class OCmp, template <class> class ToP, class Tup>
+struct DeepType<OCmp, ToP, Tup, 1> {
+  typedef typename std::tuple_element<0, Tup>::type EL1;
+  enum { value = std::is_same<OCmp, ToP<EL1>>::value };
+};
+
+template <class OCmp, template <class> class ToP, class Tup>
+struct DeepType<OCmp, ToP, Tup, 2> {
+  typedef typename std::tuple_element<0, Tup>::type EL1;
+  typedef typename std::tuple_element<1, Tup>::type EL2;
+  enum {
+    value = (std::is_same<OCmp, ToP<EL1>>::value ||
+             std::is_same<OCmp, ToP<EL2>>::value)
+  };
+};
+
+template <class OCmp, template <class> class ToP, class Tup>
+struct DeepType<OCmp, ToP, Tup, 3> {
+  typedef typename std::tuple_element<0, Tup>::type EL1;
+  typedef typename std::tuple_element<1, Tup>::type EL2;
+  typedef typename std::tuple_element<2, Tup>::type EL3;
+  enum {
+    value = (std::is_same<OCmp, ToP<EL1>>::value ||
+             std::is_same<OCmp, ToP<EL2>>::value ||
+             std::is_same<OCmp, ToP<EL3>>::value)
+  };
+};
+/*!
+ * \brief Intel CPU kernel for SpMMSum .
+ * \note it uses AVX512.
+ */
+template <class Op>
+class elem_wise_add_update : public Xbyak::CodeGenerator {
+ public:
+  typedef Op op_type;
+  typedef typename Op::type DType;
+  static_assert(
+    std::is_base_of<std::true_type, has_type<DType, supported_types>>::value,
+    "Use case fail Intel::elem_wise_add< Operator<DType> > DType is not "
+    "supported !");
+
+ protected:
+  const Xbyak::Reg64 &r_out_;
+  const Xbyak::Reg64 &r_left_;
+  const Xbyak::Reg64 &r_right;
+  const Xbyak::Reg64 &r_size_;
```

Comment:
I wonder why making them members? These members are constant, meaning they must be initialized in the constructor. And the constructor just assign them with global register names:
```c++
elem_wise_add_update()
      : r_out_(rdi),
        r_left_(rsi),
        r_right(rdx),
        r_size_(rcx),
```
What's the problem with directly using rdi/rsi/rdx/rcx? Will they be changed later?

Reply:
For better understanding the code and avoid bugs inside logic. In my opinion we should keep this form.